common/blobstore/s3store/client_test.go

```diff
@@ -0,0 +1,478 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package s3store
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/common/blobstore/blob"
+	"io/ioutil"
+	"net/url"
+	"sort"
+	"strings"
+	"testing"
+)
+
+const (
+	bucketInvalidMetadata      = "invalid-metadata"
+	bucketNoMetadata           = "bucket-no-metadata"
+	defaultBucketName          = "default-bucket-name"
+	defaultBucketOwner         = "default-bucket-owner"
+	defaultBucketRetentionDays = 10
+	customBucketNamePrefix     = "default-bucket-name-custom"
+	customBucketOwner          = "custom-bucket-owner"
+	customBucketRetentionDays  = 100
+	numberOfCustomBuckets      = 5
+)
+
+type ClientSuite struct {
+	*require.Assertions
+	suite.Suite
+}
+
+func TestClientSuite(t *testing.T) {
+	suite.Run(t, new(ClientSuite))
+}
+
+func (s *ClientSuite) SetupTest() {
+	s.Assertions = require.New(s.T())
+}
+
+func (s *ClientSuite) TestUploadDownload_Success() {
+
+	client := s.constructClient()
+	b := blob.NewBlob([]byte("body version 1"), map[string]string{})
+	key, err := blob.NewKeyFromString("blob.blob")
+	s.NoError(err)
+	s.NoError(client.Upload(context.Background(), defaultBucketName, key, b))
+	downloadBlob, err := client.Download(context.Background(), defaultBucketName, key)
+	s.NoError(err)
+	s.NotNil(downloadBlob)
+	s.assertBlobEquals(map[string]string{}, "body version 1", downloadBlob)
+	_, err = client.Delete(context.Background(), defaultBucketName, key)
```

Comment:
If you are going to call Delete in this test then lets check the blob does not exist after calling delete. Another test that is needed is to Upload/Download/Upload with same key and ensure existing blob gets replaced.